lib/web/sessions.go

```diff
@@ -122,6 +127,108 @@ func (c *SessionContext) GetClient() (auth.ClientI, error) {
 	return c.clt, nil
 }
 
+// GetUserClient will return an auth.ClientI with the role of the user at
+// the requested site. If the site is local a client with the users local role
+// is returned. If the site is remote a client with the users remote role is
+// returned.
+func (c *SessionContext) GetUserClient(site reversetunnel.RemoteSite) (auth.ClientI, error) {
+	// get the name of the current cluster
+	clt, err := c.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cn, err := clt.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// if we're trying to access the local cluster, pass back the local client.
+	if cn.GetClusterName() == site.GetName() {
+		return clt, nil
+	}
+
+	// look to see if we already have a connection to this cluster
+	remoteClt, ok := c.remoteClt[site.GetName()]
```

Comment:
this could be accessed concurrently, because the session can be used in multiple requests at the same time.

Reply:
you should protect the access to the client with read write mutex, I suggest to use separate small function with defer for that.